crates/storage/src/lazy/vec.rs

```diff
@@ -0,0 +1,689 @@
+    len: Lazy<u32, KeyType>,
+    /// The length only changes upon pushing to or popping from the vec.
+    /// Hence we can cache it to prevent unnecessary reads from storage.
+    len_cached: Option<u32>,
+    /// We use a [Mapping] to store all elements of the vector.
+    /// Each element is living in storage under `&(KeyType::KEY, index)`.
+    /// Because [StorageVec] has a [StorageKey] parameter under which the
+    /// length and element are stored, it won't collide with the other
+    /// storage fields (unless contract authors purposefully craft such a
+    /// storage layout).
+    elements: Mapping<u32, V, KeyType>,
+}
+
+impl<V, KeyType> Default for StorageVec<V, KeyType>
+where
+    V: Packed,
+    KeyType: StorageKey,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<V, KeyType> Storable for StorageVec<V, KeyType>
+where
+    V: Packed,
+    KeyType: StorageKey,
+{
+    #[inline]
+    fn encode<T: scale::Output + ?Sized>(&self, _dest: &mut T) {}
+
+    #[inline]
+    fn decode<I: scale::Input>(_input: &mut I) -> Result<Self, scale::Error> {
+        Ok(Default::default())
+    }
+
+    #[inline]
+    fn encoded_size(&self) -> usize {
+        0
+    }
+}
+
+impl<V, Key, InnerKey> StorableHint<Key> for StorageVec<V, InnerKey>
+where
+    V: Packed,
+    Key: StorageKey,
+    InnerKey: StorageKey,
+{
+    type Type = StorageVec<V, Key>;
+    type PreferredKey = InnerKey;
+}
+
+impl<V, KeyType> StorageKey for StorageVec<V, KeyType>
+where
+    V: Packed,
+    KeyType: StorageKey,
+{
+    const KEY: Key = KeyType::KEY;
+}
+
+#[cfg(feature = "std")]
+const _: () = {
+    use crate::traits::StorageLayout;
+    use ink_metadata::layout::{
+        Layout,
+        LayoutKey,
+        RootLayout,
+    };
+
+    impl<V, KeyType> StorageLayout for StorageVec<V, KeyType>
+    where
+        V: Packed + StorageLayout + scale_info::TypeInfo + 'static,
+        KeyType: StorageKey + scale_info::TypeInfo + 'static,
+    {
+        fn layout(_: &Key) -> Layout {
+            Layout::Root(RootLayout::new(
+                LayoutKey::from(&KeyType::KEY),
+                <V as StorageLayout>::layout(&KeyType::KEY),
+                scale_info::meta_type::<Self>(),
+            ))
+        }
+    }
+};
+
+impl<V, KeyType> StorageVec<V, KeyType>
+where
+    V: Packed,
+    KeyType: StorageKey,
+{
+    /// Creates a new empty `StorageVec`.
+    pub const fn new() -> Self {
+        Self {
+            len: Lazy::new(),
+            len_cached: None,
+            elements: Mapping::new(),
+        }
+    }
+
+    /// Returns the number of elements in the vector, also referred to as its length.
+    ///
+    /// The length is cached; subsequent calls (without writing to the vector) won't
+    /// trigger additional storage reads.
+    #[inline]
+    pub fn len(&self) -> u32 {
+        debug_assert!(self.len_cached.is_none() || self.len.get() == self.len_cached);
+
+        self.len_cached
```

Comment:
For that, and all other getters it would need a `&mut self` receiver.

Maybe could get around that with interior mutability to remove the inefficiency.

Reply:
Yeah I missed that. Easy to do with a `Cell` though

Reply:
@SkymanOne added it together with a test